### ISSUE TITLE
How to use Middleware in FastAPI?

### ISSUE THREAD

**Issue by hc128168** [tiangolo/fastapi#1847]
Hi,

I was trying to figure out whether FastAPI makes sense in the following application:

1. C++ Application on Server A (CA)<--middleware over TCP--> FastAPI Application on Server B <--HTTP(S)/Websocket-->mobile web browser/desktop web browser
2. C++ Application on Server B (CB)<--middleware shared memory--> FastAPI Application on Server B <--HTTP(S)/Websocket-->mobile web browser/desktop web browser

The usage cases are:

1.  Application CA and application CB want to push notifications to be displayed to all connected (logged in) browser clients.  But server A's CPU cache is very precious that the middleware has to be as light as possible and so HTTP isn't an option.  For application CB, I could find some C++ http library to interact with FastAPI application over HTTP probably but shared memory seems better as both applications live on the same server.
2.  The browser clients want to send some commands/instructions to Application CA to do something on CA.

I did read someone post a [similar question](https://github.com/tiangolo/fastapi/issues/1438).  However, the ticket was closed as the original poster didn't provide use cases.

I believe that I can replace "FastAPI Application on Server B <--HTTP(S)/Websocket-->mobile web browser/desktop web browser" with "Java Application on Server B<--TCP-->Java GUI/Android App/iOS App" in the above diagrams.

However, it seems like the development and maintenance cost are going to be quite high given I can limit a subset of browsers as the users are internal.

**Comment by pyropy** [comment]
@hc128168 Have you considered using some message queue like RabbitMQ? When your server consumes message sent from server you cant emit messages via WebSocket?

**Comment by hc128168** [comment]
@pyropy you mean something like these?

1. C++ Application on Server A (CA)<--RabbitMQ over TCP--> FastAPI Application on Server B <--HTTP(S)/Websocket-->mobile web browser/desktop web browser
2. C++ Application on Server B (CB)<--RabbitMQ's shared memory--> FastAPI Application on Server B <--HTTP(S)/Websocket-->mobile web browser/desktop web browser

**Comment by pyropy** [comment]
@hc128168 Yes something like that. You can find more info at the docs: https://www.rabbitmq.com/

But do you still have questions around how to use middleware for FastAPI? I don't see usecase for middleware to solve your problem :) 

Here's like to the FastAPI and Starlette docs regarding middlewares:

- https://www.starlette.io/middleware/
- https://fastapi.tiangolo.com/tutorial/middleware/

**Comment by hc128168** [comment]
Hi @pyropy , thanks for your suggestion and links.

I think FastAPI's and Starlette's docs define "middleware":

> A "middleware" is a function that works with every request before it is processed by any specific path operation. And also with every response before returning it.

And the middleware I referred to in this post is more like Zeromq, RabbitMQ, 29west, etc (i.e. a messaging layer).

Is there a tutorial or document somewhere that demonstrates how I can run a dedicated thread within FastAPI to process the messages from the middleware (e.g. RabbitMQ) and then push through websocket(s)?

**Comment by pyropy** [comment]
@hc128168 I didn't manage to find any tutorials that implement fastapi, rabbitmq and websocket but as from my experience you would run webserver using uvicorn or something similar and start Rabbitmq (or other messaging services) consumers running in the same event loop. 

Upon recieving the message you would emit it to whatever room or topic you choose to via websocket.

**Comment by tiangolo** [comment]
I'm not sure I understand the idea with the arrows and the shared applications, but I guess something like ZeroMQ would give you the lowest latency if that's what you are trying to optimize. Otherwise, something based on Redis (or RabbitMQ) could work.

Maybe this could help: https://github.com/encode/broadcaster

If with something like ZeroMQ, I guess you could have to run an infinite loop in an async function started at a startup event that updates the data, but not sure. :shrug:
